cypress-io/cypress
# feat: Flake Indicator ACI

- Closes #22656 , #22657 

### User facing changelog
* Display a "flaky test" indicator in the Specs Explorer for specs that are known to be flaky in the Cypress Dashboard

### Additional details
* Building new data adapter component to manage fetch/re-fetch of data for the tooltip
* Lots of cloud data management logic got added to `SpecsList` during initial ACI work - I took this opportunity to extract it out to a new composable. This helps simplify `SpecsList` so it can move back towards being a presentational component and helps organize some of the cloud spec data flows.
* Adding support for `RemoteFetchable*` types in our mock gql layer so we can leverage these in tests (like in `mountFragment`)

### Steps to test
1. Download and extract this example project: 
[flaky-e2e.zip](https://github.com/cypress-io/cypress/files/9237525/flaky-e2e.zip)
2. Open project in the app - ensure you are *logged out*
3. Verify that Specs Explorer does not display Flake Indicator, no user-noticeable changes to the UI
4. Log in to the app with your Staging Dashboard account, ACI data should populate
5. Verify that multiple specs are marked with a "Flaky" indicator (note - if other devs have recorded new runs as part of their testing the number of specs marked as flaky may change)
6. Mouse over Flaky indicator, verify tooltip opens and displays data re: flaky state of the test
    * If you want to validate the data you can review the project in the [Staging Dashboard](https://dashboard-staging.cypress.io/projects/ji3i6b/analytics/flaky-tests)
7. Verify that clicking on the indicator *or* on the tooltip content opens the dashboard to the flaky data section
8. Extra credit: Initiate a new recording from CLI, verify that the app refreshes to reflect in-progress and final data once the recording is complete (approx 30 second polling interval). Verify that flaky tooltips update with new data.

### How has the user experience changed?

https://user-images.githubusercontent.com/11482842/182251591-1e20f5d9-7b57-4e7e-a34c-0d656d055079.mov

### PR Tasks
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
